File: contracts/wasm/erc721/ts/erc721/erc721.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "wasmlib";
import * as wasmtypes from "wasmlib/wasmtypes";
import * as sc from "./index";

// Follows ERC-721 standard as closely as possible
// https://eips.ethereum.org/EIPS/eip-721
// Notable changes w.r.t. ERC-721:
// - tokenID is Hash instead of int256
// - balance amounts are Uint64 instead of int256
// - all address accounts are replaced with AgentID accounts
// - for consistency and to reduce confusion:
//     use 'approved' when it is an AgentID
//     use 'approval' when it is a Bool

// set the required base URI, to which the string encoded token ID will be concatenated
const BASE_URI = "my/special/base/uri/";
const ZERO = wasmtypes.agentIDFromBytes([]);

///////////////////////////  HELPER FUNCTIONS  ////////////////////////////

// checks if caller is owner, or one of its delegated operators
function canOperate(state: sc.MutableErc721State, caller: wasmlib.ScAgentID, owner: wasmlib.ScAgentID): bool {
    if (caller.equals(owner)) {
        return true;
    }

    let operators = state.approvedOperators().getOperators(owner);
    return operators.getBool(caller).value();
}

// checks if caller is owner, or one of its delegated operators, or approved account for tokenID
function canTransfer(state: sc.MutableErc721State, caller: wasmlib.ScAgentID, owner: wasmlib.ScAgentID, tokenID: wasmlib.ScHash): bool {
    if (canOperate(state, caller, owner)) {
        return true;
    }

    let controller = state.approvedAccounts().getAgentID(tokenID);
    return controller.value().equals(caller);
}

// common code for safeTransferFrom and transferFrom
function transfer(ctx: wasmlib.ScFuncContext, state: sc.MutableErc721State, from: wasmlib.ScAgentID, to: wasmlib.ScAgentID, tokenID: wasmlib.ScHash): void {
    let tokenOwner = state.owners().getAgentID(tokenID);
    ctx.require(tokenOwner.exists(), "tokenID does not exist");

    let owner = tokenOwner.value();
    ctx.require(canTransfer(state, ctx.caller(), owner, tokenID),
        "not owner, operator, or approved");

    ctx.require(owner.equals(from), "from is not owner");
    //TODO: ctx.require(to == <check-if-is-a-valid-address> , "invalid 'to' agentid");

    let balanceFrom = state.balances().getUint64(from);
    let balanceTo = state.balances().getUint64(to);

    balanceFrom.setValue(balanceFrom.value() - 1);
    balanceTo.setValue(balanceTo.value() + 1);

    tokenOwner.setValue(to);

    let events = new sc.Erc721Events();
    // remove approval if it exists
    let currentApproved = state.approvedAccounts().getAgentID(tokenID);
    if (currentApproved.exists()) {
        currentApproved.delete();
        events.approval(ZERO, owner, tokenID);
    }

    events.transfer(from, to, tokenID);
}

///////////////////////////  SC FUNCS  ////////////////////////////

// Gives permission to to to transfer tokenID token to another account.
// The approval is cleared when optional approval account is omitted.
// The approval will be cleared when the token is transferred.
export function funcApprove(ctx: wasmlib.ScFuncContext, f: sc.ApproveContext): void {
    let tokenID = f.params.tokenID().value();
    let tokenOwner = f.state.owners().getAgentID(tokenID);
    ctx.require(tokenOwner.exists(), "tokenID does not exist");
    let owner = tokenOwner.value();
    ctx.require(canOperate(f.state, ctx.caller(), owner), "not owner or operator");

    let approved = f.params.approved();
    if (!approved.exists()) {
        // remove approval if it exists
        let currentApproved = f.state.approvedAccounts().getAgentID(tokenID);
        if (currentApproved.exists()) {
            currentApproved.delete();
            f.events.approval(ZERO, owner, tokenID);
        }
        return;
    }

    let account = approved.value();
    ctx.require(!owner.equals(account), "approved account equals owner");

    f.state.approvedAccounts().getAgentID(tokenID).setValue(account);
    f.events.approval(account, owner, tokenID);
}

// Destroys tokenID. The approval is cleared when the token is burned.
export function funcBurn(ctx: wasmlib.ScFuncContext, f: sc.BurnContext): void {
    let tokenID = f.params.tokenID().value();
    let owner = f.state.owners().getAgentID(tokenID).value();
    ctx.require(!owner.equals(ZERO), "tokenID does not exist");
    ctx.require(ctx.caller().equals(owner), "caller is not owner");

    // remove approval if it exists
    let currentApproved = f.state.approvedAccounts().getAgentID(tokenID);
    if (currentApproved.exists()) {
        currentApproved.delete();
        f.events.approval(ZERO, owner, tokenID);
    }

    let balance = f.state.balances().getUint64(owner);
    balance.setValue(balance.value() - 1);

    f.state.owners().getAgentID(tokenID).delete();
    f.events.transfer(owner, ZERO, tokenID);
}

// Initializes the contract by setting a name and a symbol to the token collection.
export function funcInit(ctx: wasmlib.ScFuncContext, f: sc.InitContext): void {
    let name = f.params.name().value();
    let symbol = f.params.symbol().value();

    f.state.name().setValue(name);
    f.state.symbol().setValue(symbol);

    f.events.init(name, symbol);
}

// Mints tokenID and transfers it to caller as new owner.
export function funcMint(ctx: wasmlib.ScFuncContext, f: sc.MintContext): void {
    let tokenID = f.params.tokenID().value();
    let tokenOwner = f.state.owners().getAgentID(tokenID);
    ctx.require(!tokenOwner.exists(), "tokenID already minted");

    // save optional token uri
    let tokenURI = f.params.tokenURI();
    if (tokenURI.exists()) {
        f.state.tokenURIs().getString(tokenID).setValue(tokenURI.value());
    }

    let owner = ctx.caller();
    tokenOwner.setValue(owner);
    let balance = f.state.balances().getUint64(owner);
    balance.setValue(balance.value() + 1);

    f.events.transfer(ZERO, owner, tokenID);
    // if (!owner.isAddress()) {
    //     //TODO interpret to as SC address and call its onERC721Received() function
    // }
}

// Safely transfers tokenID token from from to to, checking first that contract
// recipients are aware of the ERC721 protocol to prevent tokens from being forever locked.
export function funcSafeTransferFrom(ctx: wasmlib.ScFuncContext, f: sc.SafeTransferFromContext): void {
    let from = f.params.from().value();
    let to = f.params.to().value();
    let tokenID = f.params.tokenID().value();
    transfer(ctx, f.state, from, to, tokenID);
    // if (!to.isAddress()) {
    //     //TODO interpret to as SC address and call its onERC721Received() function
    // }
}

// Approve or remove operator as an operator for the caller.
export function funcSetApprovalForAll(ctx: wasmlib.ScFuncContext, f: sc.SetApprovalForAllContext): void {
    let owner = ctx.caller();
    let operator = f.params.operator().value();
    ctx.require(!owner.equals(operator), "owner equals operator");

    let approval = f.params.approval().value();
    let approvalsByCaller = f.state.approvedOperators().getOperators(owner);
    approvalsByCaller.getBool(operator).setValue(approval);

    f.events.approvalForAll(approval, operator, owner);
}

// Transfers tokenID token from from to to.
export function funcTransferFrom(ctx: wasmlib.ScFuncContext, f: sc.TransferFromContext): void {
    let from = f.params.from().value();
    let to = f.params.to().value();
    let tokenID = f.params.tokenID().value();
    transfer(ctx, f.state, from, to, tokenID);
}

///////////////////////////  SC VIEWS  ////////////////////////////

// Returns the number of tokens in owner's account if the owner exists.
export function viewBalanceOf(ctx: wasmlib.ScViewContext, f: sc.BalanceOfContext): void {
    let owner = f.params.owner().value();
    let balance = f.state.balances().getUint64(owner);
    if (balance.exists()) {
        f.results.amount().setValue(balance.value());
    }
}

// Returns the approved account for tokenID token if there is one.
export function viewGetApproved(ctx: wasmlib.ScViewContext, f: sc.GetApprovedContext): void {
    let tokenID = f.params.tokenID().value();
    let approved = f.state.approvedAccounts().getAgentID(tokenID);
    if (approved.exists()) {
        f.results.approved().setValue(approved.value());
    }
}

// Returns if the operator is allowed to manage all the assets of owner.
export function viewIsApprovedForAll(ctx: wasmlib.ScViewContext, f: sc.IsApprovedForAllContext): void {
    let owner = f.params.owner().value();
    let operator = f.params.operator().value();
    let operators = f.state.approvedOperators().getOperators(owner);
    let approval = operators.getBool(operator);
    if (approval.exists()) {
        f.results.approval().setValue(approval.value());
    }
}

// Returns the token collection name.
export function viewName(ctx: wasmlib.ScViewContext, f: sc.NameContext): void {
    f.results.name().setValue(f.state.name().value());
}

// Returns the owner of the tokenID token if the token exists.
export function viewOwnerOf(ctx: wasmlib.ScViewContext, f: sc.OwnerOfContext): void {
    let tokenID = f.params.tokenID().value();
    let owner = f.state.owners().getAgentID(tokenID);
    if (owner.exists()) {
        f.results.owner().setValue(owner.value());
    }
}

// Returns the token collection symbol.
export function viewSymbol(ctx: wasmlib.ScViewContext, f: sc.SymbolContext): void {
    f.results.symbol().setValue(f.state.symbol().value());
}

// Returns the Uniform Resource Identifier (URI) for tokenID token if the token exists.
export function viewTokenURI(ctx: wasmlib.ScViewContext, f: sc.TokenURIContext): void {
    let tokenID = f.params.tokenID();
    if (tokenID.exists()) {
        let tokenURI = BASE_URI + tokenID.toString();
        let savedURI = f.state.tokenURIs().getString(tokenID.value());
        if (savedURI.exists()) {
            tokenURI = savedURI.value();
        }
        f.results.tokenURI().setValue(tokenURI);
    }
}

File: contracts/wasm/erc721/go/erc721/erc721.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package erc721

import "github.com/iotaledger/wasp/packages/vm/wasmlib/go/wasmlib"

// Follows ERC-721 standard as closely as possible
// https//eips.Ethereum.Org/EIPS/eip-721
// Notable changes w.R.T. ERC-721
// - tokenID is Hash instead of int256
// - balance amounts are Uint64 instead of int256
// - all address accounts are replaced with AgentID accounts
// - for consistency and to reduce confusion
//     use 'approved' when it is an AgentID
//     use 'approval' when it is a Bool

// set the required base URI, to which the base58 encoded token ID will be concatenated
const baseURI = "my/special/base/uri/"

var zero = wasmlib.ScAgentID{}

///////////////////////////  HELPER FUNCTIONS  ////////////////////////////

func approve(state MutableErc721State, owner, approved wasmlib.ScAgentID, tokenID wasmlib.ScHash) {
	state.ApprovedAccounts().GetAgentID(tokenID).SetValue(approved)
	Erc721Events{}.Approval(approved, owner, tokenID)
}

// checks if caller is owner, or one of its delegated operators
func canOperate(state MutableErc721State, caller, owner wasmlib.ScAgentID) bool {
	if caller == owner {
		return true
	}

	operators := state.ApprovedOperators().GetOperators(owner)
	return operators.GetBool(caller).Value()
}

// checks if caller is owner, or one of its delegated operators, or approved account for tokenID
func canTransfer(state MutableErc721State, caller, owner wasmlib.ScAgentID, tokenID wasmlib.ScHash) bool {
	if canOperate(state, caller, owner) {
		return true
	}

	controller := state.ApprovedAccounts().GetAgentID(tokenID)
	return controller.Value() == caller
}

// common code for safeTransferFrom and transferFrom
func transfer(ctx wasmlib.ScFuncContext, state MutableErc721State, from, to wasmlib.ScAgentID, tokenID wasmlib.ScHash) {
	tokenOwner := state.Owners().GetAgentID(tokenID)
	ctx.Require(tokenOwner.Exists(), "tokenID does not exist")

	owner := tokenOwner.Value()
	ctx.Require(canTransfer(state, ctx.Caller(), owner, tokenID),
		"not owner, operator, or approved")

	ctx.Require(owner == from, "from is not owner")
	// TODO ctx.Require(to == <check-if-is-a-valid-address> , "invalid 'to' agentid")

	nftCountFrom := state.Balances().GetUint64(from)
	nftCountTo := state.Balances().GetUint64(to)

	nftCountFrom.SetValue(nftCountFrom.Value() - 1)
	nftCountTo.SetValue(nftCountTo.Value() + 1)

	tokenOwner.SetValue(to)

	// TODO should probably clear this entry, but for now just set to zero
	approve(state, owner, zero, tokenID)

	Erc721Events{}.Transfer(from, to, tokenID)
}

///////////////////////////  SC FUNCS  ////////////////////////////

// Gives permission to to to transfer tokenID token to another account.
// The approval is cleared when the token is transferred.
func funcApprove(ctx wasmlib.ScFuncContext, f *ApproveContext) {
	tokenID := f.Params.TokenID().Value()
	tokenOwner := f.State.Owners().GetAgentID(tokenID)
	ctx.Require(tokenOwner.Exists(), "tokenID does not exist")
	owner := tokenOwner.Value()
	ctx.Require(canOperate(f.State, ctx.Caller(), owner), "not owner or operator")
	approved := f.Params.Approved().Value()
	ctx.Require(owner != approved, "approved equals owner")
	approve(f.State, owner, approved, tokenID)
}

// Destroys tokenID. The approval is cleared when the token is burned.
func funcBurn(ctx wasmlib.ScFuncContext, f *BurnContext) {
	tokenID := f.Params.TokenID().Value()
	owner := f.State.Owners().GetAgentID(tokenID).Value()
	ctx.Require(owner != zero, "tokenID does not exist")
	ctx.Require(ctx.Caller() == owner, "caller is not owner")

	approve(f.State, owner, zero, tokenID)

	balance := f.State.Balances().GetUint64(owner)
	balance.SetValue(balance.Value() - 1)
	// TODO clear this instead of setting to zero
	f.State.Owners().GetAgentID(tokenID).SetValue(zero)

	f.Events.Transfer(owner, zero, tokenID)
}

// Initializes the contract by setting a name and a symbol to the token collection.
func funcInit(ctx wasmlib.ScFuncContext, f *InitContext) {
	name := f.Params.Name().Value()
	symbol := f.Params.Symbol().Value()

	f.State.Name().SetValue(name)
	f.State.Symbol().SetValue(symbol)

	f.Events.Init(name, symbol)
}

// Mints tokenID and transfers it to caller as new owner.
func funcMint(ctx wasmlib.ScFuncContext, f *MintContext) {
	tokenID := f.Params.TokenID().Value()
	tokenOwner := f.State.Owners().GetAgentID(tokenID)
	ctx.Require(!tokenOwner.Exists(), "tokenID already minted")

	owner := ctx.Caller()
	tokenOwner.SetValue(owner)
	balance := f.State.Balances().GetUint64(owner)
	balance.SetValue(balance.Value() + 1)

	f.Events.Transfer(zero, owner, tokenID)
	//if !owner.IsAddress() {
	//	// TODO interpret to as SC address and call its onERC721Received() func
	//}
}

// Safely transfers tokenID token from from to to, checking first that contract
// recipients are aware of the ERC721 protocol to prevent tokens from being forever locked.
func funcSafeTransferFrom(ctx wasmlib.ScFuncContext, f *SafeTransferFromContext) {
	from := f.Params.From().Value()
	to := f.Params.To().Value()
	tokenID := f.Params.TokenID().Value()
	transfer(ctx, f.State, from, to, tokenID)
	//if !to.IsAddress() {
	//	// TODO interpret to as SC address and call its onERC721Received() func
	//}
}

// Approve or remove operator as an operator for the caller.
func funcSetApprovalForAll(ctx wasmlib.ScFuncContext, f *SetApprovalForAllContext) {
	owner := ctx.Caller()
	operator := f.Params.Operator().Value()
	ctx.Require(owner != operator, "owner equals operator")

	approval := f.Params.Approval().Value()
	approvalsByCaller := f.State.ApprovedOperators().GetOperators(owner)
	approvalsByCaller.GetBool(operator).SetValue(approval)

	f.Events.ApprovalForAll(approval, operator, owner)
}

// Transfers tokenID token from from to to.
func funcTransferFrom(ctx wasmlib.ScFuncContext, f *TransferFromContext) {
	from := f.Params.From().Value()
	to := f.Params.To().Value()
	tokenID := f.Params.TokenID().Value()
	transfer(ctx, f.State, from, to, tokenID)
}

///////////////////////////  SC VIEWS  ////////////////////////////

// Returns the number of tokens in owner's account if the owner exists.
func viewBalanceOf(ctx wasmlib.ScViewContext, f *BalanceOfContext) {
	owner := f.Params.Owner().Value()
	nftCount := f.State.Balances().GetUint64(owner)
	if nftCount.Exists() {
		f.Results.Amount().SetValue(nftCount.Value())
	}
}

// Returns the approved account for tokenID token if there is one.
func viewGetApproved(ctx wasmlib.ScViewContext, f *GetApprovedContext) {
	tokenID := f.Params.TokenID().Value()
	approved := f.State.ApprovedAccounts().GetAgentID(tokenID).Value()
	if approved != zero {
		f.Results.Approved().SetValue(approved)
	}
}

// Returns if the operator is allowed to manage all the assets of owner.
func viewIsApprovedForAll(ctx wasmlib.ScViewContext, f *IsApprovedForAllContext) {
	owner := f.Params.Owner().Value()
	operator := f.Params.Operator().Value()
	operators := f.State.ApprovedOperators().GetOperators(owner)
	approval := operators.GetBool(operator)
	if approval.Exists() {
		f.Results.Approval().SetValue(approval.Value())
	}
}

// Returns the token collection name.
func viewName(ctx wasmlib.ScViewContext, f *NameContext) {
	f.Results.Name().SetValue(f.State.Name().Value())
}

// Returns the owner of the tokenID token if the token exists.
func viewOwnerOf(ctx wasmlib.ScViewContext, f *OwnerOfContext) {
	tokenID := f.Params.TokenID().Value()
	owner := f.State.Owners().GetAgentID(tokenID)
	if owner.Exists() {
		f.Results.Owner().SetValue(owner.Value())
	}
}

// Returns the token collection symbol.
func viewSymbol(ctx wasmlib.ScViewContext, f *SymbolContext) {
	f.Results.Symbol().SetValue(f.State.Symbol().Value())
}

// Returns the Uniform Resource Identifier (URI) for tokenID token if the token exists.
func viewTokenURI(ctx wasmlib.ScViewContext, f *TokenURIContext) {
	tokenID := f.Params.TokenID()
	if tokenID.Exists() {
		f.Results.TokenURI().SetValue(baseURI + tokenID.String())
	}
}
